observability/logging.go

```diff
@@ -0,0 +1,157 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package observability
+
+import (
+	"fmt"
+	"strings"
+
+	"google.golang.org/grpc/binarylog"
+	binlogpb "google.golang.org/grpc/binarylog/grpc_binarylog_v1"
+	iblog "google.golang.org/grpc/internal/binarylog"
+	configpb "google.golang.org/grpc/observability/internal/config"
+	grpclogrecordpb "google.golang.org/grpc/observability/internal/logging"
+)
+
+// translateMetadata translates the metadata from Binary Logging format to
+// its GrpcLogRecord equivalent.
+func translateMetadata(m *binlogpb.Metadata) *grpclogrecordpb.GrpcLogRecord_Metadata {
+	var res grpclogrecordpb.GrpcLogRecord_Metadata
+	res.Entry = make([]*grpclogrecordpb.GrpcLogRecord_MetadataEntry, len(m.Entry))
+	for i, e := range m.Entry {
+		res.Entry[i] = &grpclogrecordpb.GrpcLogRecord_MetadataEntry{
+			Key:   e.Key,
+			Value: e.Value,
+		}
+	}
+	return &res
+}
+
+type cloudLoggingSink struct{}
+
+// Write translate a Binary Logging log entry to a GrpcLogEntry used by the gRPC
+// Observability project.
+func (cls *cloudLoggingSink) Write(binlogEntry *binlogpb.GrpcLogEntry) error {
+	var grpcLogRecord grpclogrecordpb.GrpcLogRecord
+	grpcLogRecord.Timestamp = binlogEntry.Timestamp
```

Comment:
Here and everywhere: get in the habit of using `Get`ters for fields when accessing protobufs.  What if `Write` is called with a `nil` binlogEntry?  Then this function will panic.  Maybe not possible here, but consider a helper function that is called on a sub-message:

```go
func foo(m *pb.Msg) int {
	return m.Field
}

func main() {
	var m *pb.Msg
	fmt.Println(foo(m)) // panic
}
```